File: Python/21-30/euler_23.py
from math import sqrt
import time
startTime = time.clock()
nonAbundant = [True for i in range(28124)]
total = 0
abundantNumbers = []
def sumOfFactors(number):
    result = [1]
    for i in range(1, int(sqrt(number) + 1)):
        if number % i == 0 and i != 1:
            result.append(i)
            if number / i != i:
                result.append(number / i)
    return sum(result)


for i in range(2, 28124):
    if sumOfFactors(i) > i:
        abundantNumbers.append(i)
        
for i in abundantNumbers:
    for j in abundantNumbers:
        if i + j <= 28123:
            nonAbundant[i + j] = False
        else:
            break

for i in range(0, 28124):
    if nonAbundant[i] == True:
        total = total + i
    
print total
print ("--- %s seconds ---" % (time.clock() - startTime))
    

File: Euler_C/euler_23.c
#include <stdio.h>
#include <time.h>
#include <math.h>

int sumDivisors(int input);
int result;

int main(){
  int abundant[6965] = {};
  int nonAbundant[28124] = {};
  int idx = 0;
  int sum = 0;
  clock_t start = clock();
  for(int i = 2; i < 28124; i++){ //this only works for numbers greater than 1
    if(sumDivisors(i) > i){
      abundant[idx] = i;
      idx++;
    }
  }

  for(int i = 0; i < sizeof(abundant) / sizeof(int); i++){
    for(int j = i; j < sizeof(abundant) / sizeof(int); j++){
      if(abundant[i] + abundant[j] < sizeof(nonAbundant) / sizeof(int)){
        nonAbundant[abundant[i] + abundant[j]] = 1;
      }else{
        break;
      }
    }
  }

  for(int i = 0; i < sizeof(nonAbundant) / sizeof(int); i++){
    if(!nonAbundant[i]){
      sum += i;
    }
  }

  printf("Your sum is: %d\n",  sum);
  printf("Executes in %.3fms\n", (double)(clock() - start) / CLOCKS_PER_SEC * 1000);

}

int sumDivisors(int input){
  result = 1;
  for(int i = 2; i < sqrt(input); i++){
    if(input % i == 0){
      result += i;
      result += input / i;
    }
  }
  if(sqrt(input) == (int)sqrt(input)){
    result += sqrt(input);
  }
  return result;
}
